### PR TITLE
2016 06 29 ocamlc plugins

### PR DESCRIPTION
**Pull request by lefessan**
This PR is a followup on PR #647.
It adds a new `-plugin PLUGIN` option to `ocamlc` and `ocamlopt` (and their native counterparts) to dynlink plugins in the compiler. A variable `plugin=PLUGIN` is also available in `OCAMLPARAM`.

Thanks to PR#647, plugins can be used to set new passes in the compiler (more hooks could be added later).

For that, the `Dynlink` module is copied in the compiler under the name `Compdynlink`. The code is close to what was already done for the `debugger`, which is simplified correspondingly.

Also, a new option `-bytecode` is added to `ocamldep` to only print bytecode dependencies, this is used to correctly print dependencies for `Compdynlink`, since its native and bytecode implementations are very different.

### CODE REVIEW

**Comment by nojb** [comment]
Hi @lefessan,
May I ask what kind of plugins do you have in mind ? Thanks!

**Comment by lefessan** [comment]
I have many plugins in mind, and I think other devs will have even more ideas. This feature could be used for example to experiment with compiler extensions without modifying the compiler (you would add your new optimizations or instrumentations as OPAM packages...).

**Comment by alainfrisch** [comment]
http://caml.inria.fr/mantis/view.php?id=6696

**Comment by whitequark** [comment]
:+1:

**Comment by lefessan** [comment]
If there is no objection, I will merge at noon, on Monday.

File: Makefile [hunk 1]
```diff
@@ -735,6 +736,8 @@ depend: beforedepend
 	 middle_end/base_types driver toplevel; \
 	 do $(CAMLDEP) $(DEPFLAGS) $$d/*.mli $$d/*.ml; \
 	 done) > .depend
+	$(CAMLDEP) $(DEPFLAGS) -native -impl driver/compdynlink.mlopt >> .depend
+	$(CAMLDEP) $(DEPFLAGS) -bytecode -impl driver/compdynlink.mlbyte >> .depend
```
Comment (alainfrisch) [inline Makefile]: These lines should be added to Makefile.nt as well.

File: Makefile.shared [hunk 2]
```diff
@@ -279,3 +278,35 @@ ARCHES=amd64 i386 arm arm64 power sparc s390x
 
 check_all_arches:
 	@for i in $(ARCHES); do $(MAKEREC) --no-print-directory check_arch ARCH=$$i; done
+
+# Compiler Plugins
+
+DYNLINK_DIR=otherlibs/dynlink
+
+driver/compdynlink.mlbyte: $(DYNLINK_DIR)/dynlink.ml driver/compdynlink.mli
```
Comment (alainfrisch) [inline Makefile.shared:286]: Why this dependency on compdynlink.mli?

File: driver/compplugin.ml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      if not (Hashtbl.mem plugins plugin_file) then begin
+        Compdynlink.loadfile plugin_file;
+        Hashtbl.add plugins plugin_file (); (* plugin loaded *)
+      end
+    with Compdynlink.Error error ->
+      Printf.eprintf "Error while loading plugin file %S:\n  %s\n%!"
+        plugin_file (Compdynlink.error_message error);
+      exit 2
+
+  with Compdynlink.Error error ->
```
Comment (alainfrisch) [inline driver/compplugin.ml]: Can this second handler catch anything else than the explicit `File_not_found` raise above?

In any case, the two handlers could be merged, no?

And I'd prefer to raise a different error (with a properly registered printer) rather than stop the process.

File: Makefile [hunk 4]
```diff
@@ -332,9 +332,9 @@ compilerlibs/ocamloptcomp.cma: $(MIDDLE_END) $(ASMCOMP)
 partialclean::
 	rm -f compilerlibs/ocamloptcomp.cma
 
-ocamlopt: compilerlibs/ocamlcommon.cma compilerlibs/ocamloptcomp.cma $(OPTSTART)
+ocamlopt: compilerlibs/ocamlcommon.cma compilerlibs/ocamloptcomp.cma compilerlibs/ocamlbytecomp.cma $(OPTSTART)
```
Comment (alainfrisch) [inline Makefile:338]: Note for the reader: we need to ocamlbytecomp.cma in ocamlopt because bytecode Dynlink relies on many modules from this library.

It's a bit annoying, but it's probably not enough of a problem to deserve a more complex solution.

File: Makefile [hunk 5]
```diff
@@ -381,6 +381,7 @@ partialclean::
 	rm -f compilerlibs/ocamlopttoplevel.cmxa
 
 ocamlnat: compilerlibs/ocamlcommon.cmxa compilerlibs/ocamloptcomp.cmxa \
+    compilerlibs/ocamlbytecomp.cmxa \
```
Comment (alainfrisch) [inline Makefile:387]: Same question as above.  And also (if ocamlbytecomp is indeed required), you should udate Makefile.nt as well.

File: toplevel/opttopdirs.ml [hunk 6]
```diff
@@ -113,7 +113,7 @@ type 'a printer_type_old = 'a -> unit
 let match_printer_type ppf desc typename =
   let printer_type =
     try
-      Env.lookup_type (Ldot(Lident "Topdirs", typename)) !toplevel_env
+      Env.lookup_type (Ldot(Lident "Opttopdirs", typename)) !toplevel_env
```
Comment (alainfrisch) [inline toplevel/opttopdirs.ml:116]: Can you confirm this is an unrelated bug fix?

**Comment by alainfrisch** [comment]
I'm generally in favor of this change.  One aspect where it is a bit dangerous is that it will encourage people to rely on dynlink, even though it might not be available on some platform (at least for the native compiler).  This means that published packages will tend to be broken (or less usable) on those platforms more often than today.  Workarounds will be either to use the bytecode version of the compilers (ocamlc/ocamlopt), and fixing packages so that they install .cma in addition to .cmxs if they don't already; or to resort to statically linking custom compiler drivers embedding the plugins.

File: Makefile.shared (hunk 2 above)
Comment (lefessan) [inline Makefile.shared:286]: To force the creation of `driver/compdynlink.mli` at the same time

File: toplevel/opttopdirs.ml (hunk 6 above)
Comment (lefessan) [inline toplevel/opttopdirs.ml:116]: Yes, it fixes a problem I had while testing that `ocamlnat` was still working after my changes.

File: Makefile (hunk 5 above)
Comment (lefessan) [inline Makefile:387]: The rule for `ocamlnat` is very different in `Makefile.nt`, so different, actually, that I am not very confident it is still correct (I couldn't find any occurrence of the definition of the `NATTOPOBJS` variable used there)

File: Makefile (hunk 1 above)
Comment (lefessan) [inline Makefile]: Done.

File: driver/compplugin.ml (hunk 3 above)
Comment (lefessan) [inline driver/compplugin.ml]: Indeed, the two handlers were redondant.

> And I'd prefer to raise a different error (with a properly registered printer) rather than stop the process.

Done.

File: Makefile (hunk 5 above)
Comment (lefessan) [inline Makefile:387]: The second commit of this PR is an experiment: `Compdynlink` is used instead of `Dynlink` in `toplevel/*.ml`, so removing the dependency on `dynlink.cmxa` for `ocamlnat`.

Does that make sense ? I tested it and it works, but I couldn't test it on Windows (I am on holidays).

**Comment by lefessan** [comment]
The patch was reviewed, and there was no objection, so merging.

**Comment by alainfrisch** [comment]
Can you update the changelog, the man page and the manual?

**Comment by Kakadu** [comment]
Also, it will be great to have a helloworld plugin somewhere too.

**Comment by whitequark** [comment]
@Kakadu I believe this only introduces the loading mechanism, no new extensibility is added to the compiler. Of course you can still mutate the global data exposed by compiler-libs but that does not go very far.

**Comment by let-def** [comment]
Link to the updated changelog, man page and manual: https://github.com/ocaml/ocaml/pull/698

**Comment by Kakadu** [comment]
Thanks, @let-def 

Do you know where can I found a demo  hello-world plugin? Should this feature deprecate typed-ppx ?

**Comment by let-def** [comment]
No idea.

**Comment by lefessan** [comment]
Since it seems there is some interest in ocamlc/ocamlopt plugins, here are a few examples:

https://github.com/OCamlPro/ocamlc-plugins

Note that, with 4.04.0, ocamlopt.opt cannot load plugins, I fixed it by recreating a correct `ocamlopt.opt` in `ocamlc-plugins`, and by submitting a PR for ocaml to fix the problem upstream:

https://github.com/ocaml/ocaml/pull/980

**Comment by Kakadu** [comment]
@lefessan Do you think that it is a good idea to add some hooks for typed tree in the same manner as typed PPX do? I mean: compiler typechecks, plugin does some stuff, compiler untypes, compiler typechecks, etc....

**Comment by lefessan** [comment]
@Kakadu I updated ocamlc-plugins with an example `pptx_example` of rewriter on the Typedtree.

**Comment by Kakadu** [comment]
@lefessan Your typed tree example seems to have a limited functionality. We will now the type name of a value but we can't get information about this type because (it seems for me) we do not pass environment to a plugin.

I expected to see something like that: plugins does some stuff, untypes, than compiler types the parsetree again and sends to the next plugin.  It seems possible to implement a one more typechecking phase inside a plugin itself but not without environment.

**Comment by lefessan** [comment]
@kakadu Check `typedtree.ml`. All expressions carry both the type of the expression, and the full enviroment.

For untyping and retyping, it is clearly not possible at this point. That would need a bigger change in the compiler pipeline.

**Comment by Kakadu** [comment]
@lefessan thanks
